### PR TITLE
fix(state-sync): Make partial chunks message deterministic

### PR DESCRIPTION
**Pull request by nikurt**
The issue only happens if a node tracks a subset of shards.
The order of shards is arbitrary because:

* Shard ids are in a HashSet
* In one case the first the node adds the shards that are cached, and later the shards that are only available on disk.

### CODE REVIEW

File: core/primitives/src/sharding.rs [hunk 1]
```diff
@@ -601,6 +602,19 @@ pub struct ShardProof {
...
+impl PartialOrd<Self> for ReceiptProof {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ReceiptProof {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.1.from_shard_id, self.1.to_shard_id)
+            .cmp(&(other.1.from_shard_id, other.1.to_shard_id))
```
Comment (wacban) [inline core/primitives/src/sharding.rs:614]: This doesn't seem sufficiently ordered, there can be multiple receitpts with the same from and to shard ids and thus still be non-deterministic. How about you also compare the hashes?

**Review by wacban — COMMENTED** [review]
Rather than sorting here and there can you use a sorted data structure?

File: core/primitives/src/sharding.rs (hunk 1 above)
Comment (nikurt) [reply core/primitives/src/sharding.rs:614]: The receipts are grouped by shard ids of outgoing and incoming:

```
pub struct ReceiptProof(pub Vec<Receipt>, pub ShardProof);
```

Comment (nikurt) [reply core/primitives/src/sharding.rs:614]: @wacban or is your suggestion to order the receipts inside `ReceiptProof`? :thinking:

Comment (wacban) [reply core/primitives/src/sharding.rs:616]: I see now what's going on. You're claiming that there is at most one ReceiptProof per every pair (from, to) and as such the comparison is actually strict. As long as the assumption is correct I'm fine with this implementation. 

I was not suggesting sorting the receipts but that doesn't sound too bad just to be extra cautious. I lack context to say if it is actually needed so I'll leave it up to your best judgment.

File: pytest/tests/sanity/state_sync_then_catchup.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env python3
```
Comment (wacban) [inline pytest/tests/sanity/state_sync_then_catchup.py:1]: nit: It's good practice to wrap the code in a unittest and a _ _ main _ _ like so:
https://github.com/near/nearcore/tree/master/pytest#creating-new-tests

File: pytest/tests/sanity/state_sync_then_catchup.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+result = boot_node.send_tx_and_wait(deploy_contract_tx, 10)
+assert 'result' in result and 'error' not in result, (
+    'Expected "result" and no "error" in response, got: {}'.format(result))
+
+
+def epoch_height(block_height):
+    if block_height == 0:
+        return 0
+    if block_height <= EPOCH_LENGTH:
+        # According to the protocol specifications, there are two epochs with height 1.
```
Comment (wacban) [inline pytest/tests/sanity/state_sync_then_catchup.py:118]: TIL & WTF

File: pytest/tests/sanity/state_sync_then_catchup.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            payment_tx = transaction.sign_payment_tx(key_from, account_to, 1,
+                                                     nonce, last_block_hash)
+            boot_node.send_tx(payment_tx).get('result')
+        else:
+            key = random_u64()
+            keys.append(key)
+            call_function('write', key, nonce, boot_node.signer_key,
+                          last_block_hash)
+            call_function('write', key, nonce, node1.signer_key,
+                          last_block_hash)
```
Comment (wacban) [inline pytest/tests/sanity/state_sync_then_catchup.py:162]: Can you add a small comment on what exactly are you doing here? 
Also nit: early returns may flatten this code a bit.

File: pytest/tests/sanity/state_sync_then_catchup.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+# Run node0 more to trigger block sync in node1.
+nonce, keys = random_workload_until(int(EPOCH_LENGTH * 3), nonce, keys,
+                                    boot_node)
+
+# Node1 is now behind and needs to do header sync and block sync.
+node1.start(boot_node=boot_node)
+node1_height = node1.get_latest_block().height
+logger.info(f'started node1@{node1_height}')
+
+nonce, keys = random_workload_until(int(EPOCH_LENGTH * 3.7), nonce, keys, node1)
```
Comment (wacban) [inline pytest/tests/sanity/state_sync_then_catchup.py:200]: Are there any checks that you also need to do here or does the workload do it?

File: pytest/tests/sanity/state_sync_then_catchup.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+import transaction
+import utils
+
+from configured_logger import logger
+
+EPOCH_LENGTH = 50
+
+state_parts_dir = str(pathlib.Path(tempfile.gettempdir()) / 'state_parts')
+
+config0 = {
```
[the remark this answers is not in the document]
Comment (nikurt) [reply pytest/tests/sanity/state_sync_then_catchup.py:30]: More than half of the fields are different or have different values

File: pytest/tests/sanity/state_sync_then_catchup.py (hunk 2 above)
Comment (nikurt) [reply pytest/tests/sanity/state_sync_then_catchup.py:1]: Added `main()`, thanks!
Nayduck doesn't expect these tests to be unittests.

File: pytest/tests/sanity/state_sync_then_catchup.py (hunk 5 above)
Comment (nikurt) [reply pytest/tests/sanity/state_sync_then_catchup.py:200]: No checks needed. Without the fix the node simply crashes and doesn't reach the target height.

File: chain/chain/src/resharding.rs [hunk 7]
```diff
@@ -362,7 +362,7 @@ mod tests {
             state_root = new_state_root;
 
             // update split states
-            let trie_changes = tries
+            let mut trie_updates = tries
```
Comment (wacban) [inline chain/chain/src/resharding.rs:365]: Sus, this seems a bit similar to changes from a PR by Shreyan...

File: nearcore/src/runtime/mod.rs [hunk 8]
```diff
@@ -1076,28 +1076,40 @@ impl RuntimeAdapter for NightshadeRuntime {
         block_hash: &CryptoHash,
         state_roots: HashMap<ShardUId, StateRoot>,
         next_epoch_shard_layout: &ShardLayout,
-        state_changes: StateChangesForSplitStates,
+        state_changes_for_split_states: StateChangesForSplitStates,
```
Comment (wacban) [inline nearcore/src/runtime/mod.rs:1079]: sus sus sus

File: pytest/tests/sanity/state_sync_then_catchup.py (hunk 2 above)
Comment (wacban) [reply pytest/tests/sanity/state_sync_then_catchup.py:1]: I think as long as you run unittest.run() from within main it should work fine. Did you try and did it not work? Maybe you also need to put the test logic in a method with name beginning with test_. Not a biggie though, don't worry about it too much.

**Review by wacban — APPROVED** [review]
approving as looks good but you might have mixed up some extra code from other pr, please review carefully

File: chain/chain/src/resharding.rs (hunk 7 above)
Comment (nikurt) [reply chain/chain/src/resharding.rs]: Didn't touch that file. Rebased and this change disappeared.

File: nearcore/src/runtime/mod.rs (hunk 8 above)
Comment (nikurt) [reply nearcore/src/runtime/mod.rs]: Didn't touch that file. Rebased and this change disappeared.

**Pull request merged**